solr/solrj/src/java/org/apache/solr/client/solrj/io/Tuple.java

```diff
@@ -87,10 +87,20 @@ public Tuple(String k1, Object v1, String k2, Object v2) {
    * @param fields map containing keys and values to be copied to this tuple
    */
   public Tuple(Map<String, ?> fields) {
-    // TODO Use bulk putAll operation that will properly size the map
-    // https://issues.apache.org/jira/browse/SOLR-15480
-    for (Map.Entry<String, ?> entry : fields.entrySet()) {
-      put(entry.getKey(), entry.getValue());
+    putAll(fields);
+  }
+
+  /**
+   * A copy constructor
+   * @param original Tuple that will be copied
+   */
+  public Tuple(Tuple original) {
+    this.putAll(original.fields);
+    if (original.fieldNames != null) {
+      this.fieldNames = new ArrayList<>(original.fieldNames);
+    }
+    if (original.fieldLabels != null) {
+      this.fieldLabels = new HashMap<>(original.fieldLabels);
```

Comment:
I agree. I think ensuring the implementation type of the Map stays the same, especially since `Tuple.fieldLabels` is denoted as `Map<String, String>` to the rest of the code base, would probably be the expectation for the caller. The same could be said for `Tuple.fieldNames` as well since its written as `List<String>`, so a caller may not expect their implementation type to change either.

The trouble I'm running into currently is that the implementation classes `HashMap` and `ArrayList` both implement `Cloneable` and provide a public `clone` method. However, the `Map` and `List` interfaces do not and they are the only ones that, from what I can tell, are the only types we could reason about. My initial though is to do some kind of reflection to get the constructors of the given types and use those to instantiate he clones, but that doesn't feel quite right to me.

```
public Tuple(Tuple original) {
    this.putAll(original.fields);
    if (original.fieldNames != null) {
      final Class<? extends List> implementationClass = original.fieldNames.getClass();
      this.fieldNames = (List<String>) implementationClass.getDeclaredConstructor().newInstance(original.fieldNames);
    }
    if (original.fieldLabels != null) {
      final Class<? extends Map> implementationClass = original.fieldLabels.getClass();
      this.fieldLabels = (Map<String, String>) implementationClass.getDeclaredConstructor().newInstance(original.fieldLabels);
    }
  }
```

Does that seem like the right approach, or would there be a better way to handle this?

Reply:
Btw, this led me to do some reading up on the `Cloneable` interface online and in Effective Java, and it is interesting :smile:

Reply:
> ... The same could be said for `Tuple.fieldNames` as well since its written as `List<String>`, so a caller may not expect their implementation type to change either. ...

Good point.

> ... My initial thought is to do some kind of reflection to get the constructors of the given types and use those to instantiate the clones, but that doesn't seem quite right to me. ... Does that seem like the right approach, or would there be a better way to handle this?

Thanks for looking into! Good questions. I share your "doesn't seem quite right" sentiment.

Okay, maybe taking a step back and/or looking slightly sideways at things might move us forward here, thinking out aloud:

* What does it mean to "merge" one tuple into another?
  * Is it meant to be the fields only (current implementation) or should field labels and names be included?
    * What if there is overlap between the two field labels lists? Should the resulting list have duplicates or should only not-yet-present elements be added?
    * What if there is overlap between the two field names maps? Does the caller intend for the "source" tuple (`other`) to override the content of the "destination" tuple (`this`) or should the "source" prevail or should some sort of "renaming" happen?
  * If (and that is an if) field labels and names are _not_ to be included then a "fix" could be simply to document that is the intended behaviour and anyone who needs some other behaviour, well, they may call the setter methods for the field labels and field names.
  * If the "merge" method code did not change then the question of field labels and field names container type preservation does not arise for it (in the scenario where the "source" tuple does not have a container as yet).
  * If the "merge" method is documented as having only `putAll` behaviour then one could also argue that callers might as well use `putAll` directly i.e. deprecation and eventual removal of "merge" as a way to "fix" it.
  
* What does it mean to "clone" a tuple?
  * Cloning including field labels and field names, that seems uncontroversial.
  * Cloning preserving field label and field names container type, or not preserving it, that is the question.  

* The use of field labels and field names seems relatively rare.
  * To help with code base reading I've opened #256 and #257 to remove deprecated things.

Reply:
> ... The use of field labels and field names seems relatively rare. ...

@joel-bernstein - would love to hear your insights and thoughts on this? thanks!

Reply:
> ... What if there is overlap between the two field names lists? ...
> ... What if there is overlap between the two field labels maps? ...

> ... The use of field labels and field names seems relatively rare. ...

So it seems there's currently no `getFieldLabels()` or `getFieldNames()` use in the code base and looking at the use within the `Tuple.writeMap` method I think we don't need to worry about hypothetically a sorted map turning into an unsorted map via merging or cloning i.e. we can go with the `ArrayList` and `HashMap` as it is, sorry for the distraction and diversion into cloning and reflection territory ...

Interestingly though the `Tuple.writeMap` method uses `EntryWriter.put` that mentions to not call it twice with the same key -- https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.9.0/solr/solrj/src/java/org/apache/solr/common/MapWriter.java#L87 -- and so the overlap when merging question seems relevant, there's no `TupleTest` class as yet but maybe there's an opportunity here to start one?

Reply:
No problem at all! I wasn't sure if you wanted to talk it over with someone else before we moved forward, so I decided to take a backseat for a bit.

The original `merge` method just did a `putAll`, so we could think of the behavior of the method still following the same principle. When I look at how `Map.put` or `Map.putAll` method works, it will overwrite the old value with the new. `Map` also has a merge method that takes a `BiFunction` with a key and value, but that may be overkill unless we wanted a separate `merge` method that took at `BiFunction` as well to allow flexibility. So, using the idea of the original "putAll merge", then I would think we just overwrite in such a way that `Tuple other`'s `field`s, `fieldNames`, and `fieldLabels` have higher precedence over the originals.

For example:
if we have a `Tuple original` where `field = 'hello'` with an associated `fieldName` and `fieldLabel` entry, and we have a `Tuple other` with just a `field = 'hello'` and no associated `fieldName`/`fieldLabel`, then we take `other`'s state.

Which sounds like something that really needs a test! I'll start writing those.

Reply:
After taking some vacation time, I looked back at this, and I think my thought process above over complicated this. Instead I think we should do the following:
1. Give `other` precedence and do `putAll`s for `fields` and `fieldLabels`.
2. Add individual `fieldName` items from `other` only if they aren't present in `this.fieldNames`.

Reply:
I thought about the case where we rename `other.fieldNames` on collision, but I'm not sure where we would call them since the only real "identifier" that `Tuple`'s would have would be either their position (this || other) or there java object reference.

Reply:
> 1. Give `other` precedence and do `putAll`s for `fields` and `fieldLabels`.
> 2. Add individual `fieldName` items from `other` only if they aren't present in `this.fieldNames`.

Yes, that makes sense to me too. Looking at how `clone` and `merge` are used e.g. `HashJoinStream` there is documented also -- e.g. https://solr.apache.org/guide/8_9/stream-decorator-reference.html#hashjoin -- w.r.t. right being used if both left and right contain the same field. So `other` being given precedence would be consistent with that, assuming `left.merge(right)` directionality.

> The only downsides I could see would be if the user had a `field` that they didn't want externally serializable (ie: it doesn't have an associated `fieldName/fieldLabel`), then `other` could override that original state. That may be fine in this case. If a user wants more fine grain control, then they should probably just do the merge "by hand" instead of relying on `Tuple.merge`.

Yes, that's one of the edge cases where `tupleA` and `tupleB` could have labels and names for a particular purpose but the merged result might not necessarily suit the purpose. As you say, in that scenario alternative "merging" could be done, or regular `Tuple.merge` could be called followed by some sort of post-processing.

Another edge case is collisions in the field labels, I'll add a `TupleTest.writeMapTest()` for that, let me know what you think and/or feel free to amend or revert.

Reply:
> I thought about the case where we rename `other.fieldNames` on collision as well, but I'm not sure where we would call them since the only real "identifier" that `Tuple`'s would have would be either their position (this || other) or their java object reference.

My interpretation of `fieldNames` is as a filter of what is to be serialised, possibly with an ordering also, i.e. no collision as such. And I agree that there aren't any real "identifiers" that might be used in renaming. And `fieldNames` and `fieldLabels` being used together, would renaming in the former then also need to imply renaming in the latter?

(It's not very "merge like" but in principle collision or overlap could also be resolved via removal: `[A B C] merged with [B  D F] resulting in [A C D F]` with the common `B` removed but that seems quite counter-intuitive not least how would the caller know that `B` was removed?)

Reply:
That's how I interpreted `fieldNames` too. If a user starts using them along with `fieldLabels` then they probably want more control in regards to how they serialize things.

> would renaming in the former then also need to imply renaming in the latter?  

It could, I didn't think about that. Also..

> (It's not very "merge like" but in principle collision or overlap could also be resolved via removal: [A B C] merged with [B D F] resulting in [A C D F] with the common B removed but that seems quite counter-intuitive not least how would the caller know that B was removed?)

that's an interesting idea too. However, I think if a developer wasn't aware of the removal property, then it could be a bit of a surprise. I think in either the removal case or the renaming case, it would create more work for a developer versus just one method call to `Tuple.merge` ...maybe... 

I think our idea of just having `Tuple.merge` do what we agreed on above and having a user take matters into their own hands if they wanted a different merge result would be the best offering.

Reply:
> Another edge case is collisions in the field labels, I'll add a TupleTest.writeMapTest() for that, let me know what you think and/or feel free to amend or revert.

I think the the test your wrote looks good! I think it clearly illustrates usage of Tuple. I went back and update the `TupleTest.mergeTest` to where the `fieldNames` and `fieldLabels` relate to the `fields` key to go along with the theme of having the tests mimic actual use cases.